control/prometheus.py

```diff
@@ -0,0 +1,164 @@
+#
+#  Copyright (c) 2021 International Business Machines
+#  All rights reserved.
+#
+#  SPDX-License-Identifier: LGPL-3.0-or-later
+#
+#  Authors: pcuzner@ibm.com
+#
+
+import os
+import logging
+import spdk.rpc as rpc
+
+from prometheus_client.core import REGISTRY, GaugeMetricFamily, CounterMetricFamily
+from prometheus_client import start_http_server
+
+
+def start_exporter(spdk_rpc_client, port):
+    """Start the prometheus exporter and register the NVMeOF collector"""
+    start_http_server(port)
+    REGISTRY.register(NVMeOFCollector(spdk_rpc_client))
+
+
+class NVMeOFCollector:
+    """Provide a prometheus endpoint for nvmeof gateway statistics"""
+
+    def __init__(self, spdk_rpc_client):
+        self.logger = logging.getLogger(__name__)
+        self.spdk_rpc_client = spdk_rpc_client
+        self.metric_prefix = "ceph_nvmeof"
+
+    def collect(self):
```

Comment:
TBH, I'm not a fan of a TTL for a prometheus scrape, since the scrapes are at intervals and if you hit the cache the data won't necessarily be at least 'interval' secs old and there's no link between the ttl hardcoded limit in the code and the prometheus jobs scrape_interval parameter.

I tested up to 2048 bdevs, and didn't hit any drop outs.

Even in the context of multi-cluster, it's the API that is hit not the endpoint...right?

Reply:
Yup, obviously the problem will always come from the underlying stats API, but, as usual too, we often don't have much room to improve that... That's why TTL caching was also introduced in the Ceph exporter.

Additionally, while regular/well-intentioned usage may not be problematic, if this is an unauthenticated endpoint (as Prometheus ones are), anyone could abuse that to trigger a DoS attack. That's why in general I find it useful to put some throttling mechanism that ensures that a peak in the rate of requests doesn't inevitably translate into a service loss.

Reply:
The latest version places the collection into a separate thread, and includes timings for all method calls in the prometheus payload.